frontend/catalyst/compiler.py

```diff
@@ -174,12 +174,13 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
         "quantum-bufferize",
         "func-bufferize",
         "func.func(finalizing-bufferize)",
+        "canonicalize",
         "func.func(buffer-hoisting)",
         "func.func(buffer-loop-hoisting)",
         "func.func(buffer-deallocation)",
         "convert-arraylist-to-memref",
-        "canonicalize",
         "convert-bufferization-to-memref",
+        "canonicalize",
```

Comment:
We canonicalize twice because canonicalize is one of the passes that is able to remove some unused instructions. In particular memref_to_tensor or tensor_to_memref. For the Callback lowering passes, we needed to move canonicalize slightly earlier. But Sergei found that calling canonicalize at that specific position creates issues for this new PR. So the solution was to canonicalize early enough to eliminate memref_to_tensor / tensor_to_memref but late enough to avoid Sergei's issue.

Reply:
Ok thanks, could you add a comment about that?

Reply:
I've left review comments here. https://github.com/PennyLaneAI/catalyst/pull/890#discussion_r1662999004 and here https://github.com/PennyLaneAI/catalyst/pull/890#discussion_r1663001260